vermouth/gmx/itp_read.py

```diff
@@ -0,0 +1,533 @@
+    def _block_atoms(self, line, lineno=0):
+        tokens = collections.deque(_tokenize(line))
+        _parse_block_atom(tokens, self.current_block)
+
+    @SectionLineParser.section_parser('moleculetype', 'bonds', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'angles', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'dihedrals', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'impropers', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'constraints', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'pairs', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'exclusions', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'virtual_sites2', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'position_restraints', context_type='block')
+
+    def _interactions(self, line, lineno=0, context_type=''):
+
+        context = self.current_block
+        interaction_name = self.section[-1]
+        delete = False
+
+        tokens = collections.deque(_tokenize(line))
+
+        n_atoms = self.interactions_natoms.get(interaction_name)
+
+        _base_parser(
+            tokens,
+            context,
+            context_type=context_type,
+            section=interaction_name,
+            natoms=n_atoms,
+            delete=delete,
+            current_meta=self.current_meta
+            )
+
+
+    @SectionLineParser.section_parser('moleculetype', 'virtual_sitesn', context_type='block')
+    def _vsn_interactions(self, line, lineno=0, context_type=''):
+
+        context = self.current_block
+        interaction_name = self.section[-1]
+        delete = False
+
+        tokens = collections.deque(_tokenize(line))
+
+        n_atoms = self.interactions_natoms.get(interaction_name)
+
+        _base_parser(
+            tokens,
+            context,
+            context_type=context_type,
+            section=interaction_name,
+            natoms=n_atoms,
+            delete=delete,
+            current_meta=self.current_meta
+            )
+   
+        # we need to reorder the atoms and interactions, because for virtual_sitesn
+        # the function type is on section position followed by an undefined number
+        # of atom indices from which the VS is built
+ 
+        if '--' not in line:
+           first_atom = context.interactions['virtual_sitesn'][-1].atoms[0]
+           other_atoms = context.interactions['virtual_sitesn'][-1].atoms[2:]
+           atoms = [first_atom] + other_atoms
+   
+           sec_atom = context.interactions['virtual_sitesn'][-1].atoms[1]
+           params = [sec_atom]+context.interactions['virtual_sitesn'][-1].parameters
+       
+           context.interactions['virtual_sitesn'][-1].atoms[:]=atoms
+           context.interactions['virtual_sitesn'][-1].parameters[:]=params
```

Comment:
Since you do the reshuffling *after* adding the interaction, could you try this with e.g.
```
[ atoms ]
1 ...
2 ...
3 ...

[ virtuals_sitesn ]
1 10 2 3
```

Reply:
yes this does indeed fail :/

Reply:
new implementation can parse VS as regular interaction